File: src/main/java/rails/game/action/ReachDestinations.java
package rails.game.action;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.util.ArrayList;
import java.util.List;

import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;

import com.google.common.base.Objects;

import net.sf.rails.game.CompanyManager;
import net.sf.rails.game.PublicCompany;
import net.sf.rails.game.RailsRoot;
import net.sf.rails.util.RailsObjects;
import net.sf.rails.util.Util;


/** This class is needed until we have a means to determine reaching
 * destinations automatically.
 *
 * Rails 2.0: updated equals and toString methods
 */
public class ReachDestinations extends PossibleORAction {

    // Server-side settings
    transient protected List<PublicCompany> possibleCompanies;
    protected String possibleCompanyNames;

    // Client-side settings
    transient protected List<PublicCompany> reachedCompanies;
    protected String reachedCompanyNames = "";

    public static final long serialVersionUID = 1L;

    public ReachDestinations (RailsRoot root, @NotNull List<PublicCompany> possibleCompanies) {
        super(root);
        this.possibleCompanies = possibleCompanies;
        StringBuilder b = new StringBuilder();
        for (PublicCompany company : possibleCompanies) {
            if (b.length() > 0) b.append(",");
            b.append (company.getId());
        }
        possibleCompanyNames = b.toString();
    }

    public List<PublicCompany> getPossibleCompanies() {
        return possibleCompanies;
    }

    public String getPossibleCompanyNames () {
        return possibleCompanyNames;
    }

    public void addReachedCompany (PublicCompany company) {
        reachedCompanies.add (company);
        if (reachedCompanyNames.length() > 0) {
            reachedCompanyNames += ",";
        }
        reachedCompanyNames += company.getId();
    }

    public List<PublicCompany> getReachedCompanies() {
        return reachedCompanies;
    }

    @Override
    protected boolean equalsAs(PossibleAction pa, boolean asOption) {
        // identity always true
        if (pa == this) return true;
        //  super checks both class identity and super class attributes
        if (!super.equalsAs(pa, asOption)) return false;

        // check asOption attributes
        ReachDestinations action = (ReachDestinations)pa;
        boolean options = Objects.equal(this.possibleCompanies, action.possibleCompanies);

        // finish if asOptions check
        if (asOption) return options;

        // check asAction attributes
        return options && Objects.equal(this.reachedCompanies, action.reachedCompanies);
    }

    @Override
    public String toString() {
        return super.toString() +
                RailsObjects.stringHelper(this)
                    .addToString("possibleCompanies", possibleCompanies)
                    .addToStringOnlyActed("reachedCompanies", reachedCompanies)
                    .toString();
    }

    private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
        in.defaultReadObject();

        CompanyManager cmgr = getCompanyManager();

        possibleCompanies = new ArrayList<>();
        if (Util.hasValue(possibleCompanyNames)) {
            for (String cname : possibleCompanyNames.split(",")) {
                if ( StringUtils.isNotBlank(cname) ) {
                    possibleCompanies.add(cmgr.getPublicCompany(cname));
                }
            }
        }

        if (Util.hasValue(reachedCompanyNames)) {
            reachedCompanies = new ArrayList<>();
            for (String cname : reachedCompanyNames.split(",")) {
                if ( StringUtils.isNotBlank(cname) ) {
                    reachedCompanies.add(cmgr.getPublicCompany(cname));
                }
            }
        }
    }
}

File: dotnet/RailsLib.Net/Rails/Game/Action/ReachDestinations.cs
using GameLib.Net.Game;
using GameLib.Net.Util;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;

/** This class is needed until we have a means to determine reaching
 * destinations automatically.
 *
 * Rails 2.0: updated equals and toString methods
 */

namespace GameLib.Rails.Game.Action
{
    [JsonObject(MemberSerialization.Fields)]
    public class ReachDestinations : PossibleORAction
    {
        // Server-side settings
        /*transient*/
        [JsonIgnore]
        protected List<PublicCompany> possibleCompanies;
        protected string possibleCompanyNames = "";

        // Client-side settings
        /*transient*/
        [JsonIgnore]
        protected List<PublicCompany> reachedCompanies;
        protected string reachedCompanyNames = "";

        new public const long serialVersionUID = 1L;

        public ReachDestinations(List<PublicCompany> companies)
        {
            possibleCompanies = companies;
            StringBuilder b = new StringBuilder();
            foreach (PublicCompany company in companies)
            {
                if (b.Length > 0) b.Append(",");
                b.Append(company.Id);
            }
            possibleCompanyNames = b.ToString();
        }

        /** Required for deserialization */
        public ReachDestinations() { }

        public List<PublicCompany> PossibleCompanies
        {
            get
            {
                return possibleCompanies;
            }
        }

        public string PossibleCompanyNames
        {
            get
            {
                return possibleCompanyNames;
            }
        }

        public void AddReachedCompany(PublicCompany company)
        {
            if (reachedCompanies == null)
                reachedCompanies = new List<PublicCompany>();
            reachedCompanies.Add(company);
            if (reachedCompanyNames.Length > 0)
            {
                reachedCompanyNames += ",";
            }
            reachedCompanyNames += company.Id;
        }

        public List<PublicCompany> ReachedCompanies
        {
            get
            {
                return reachedCompanies;
            }
        }

        override protected bool EqualsAs(PossibleAction pa, bool asOption)
        {
            // identity always true
            if (pa == this) return true;
            //  super checks both class identity and super class attributes
            if (!base.EqualsAs(pa, asOption)) return false;

            // check asOption attributes
            ReachDestinations action = (ReachDestinations)pa;
            // #FIXME sequence compare
            bool options = possibleCompanies.SequenceEqual(action.possibleCompanies);

            // finish if asOptions check
            if (asOption) return options;

            // check asAction attributes
            return options && reachedCompanies.SequenceEqual(action.reachedCompanies);
        }

        override public string ToString()
        {
            return base.ToString() +
                    RailsObjects.GetStringHelper(this)
                        .AddToString("possibleCompanies", possibleCompanies)
                        .AddToStringOnlyActed("reachedCompanies", reachedCompanies)
                        .ToString();
        }

        [OnDeserialized]
        new internal void OnDeserialized(StreamingContext context)
        {
            CompanyManager cmgr = CompanyManager;

            possibleCompanies = new List<PublicCompany>();
            if (!string.IsNullOrEmpty(possibleCompanyNames))
            {
                foreach (string cname in possibleCompanyNames.Split(','))
                {
                    if (!string.IsNullOrEmpty(cname))
                    {
                        possibleCompanies.Add(cmgr.GetPublicCompany(cname));
                    }
                }
            }
            reachedCompanies = new List<PublicCompany>();
            if (!string.IsNullOrEmpty(reachedCompanyNames))
            {
                foreach (string cname in reachedCompanyNames.Split(','))
                {
                    if (!string.IsNullOrEmpty(cname))
                    {
                        reachedCompanies.Add(cmgr.GetPublicCompany(cname));
                    }
                }
            }
        }
    }
}
